feat: create a simple Sign API example on top of the WS client

silvestrst-crypto opened the pull request:
# Description

This change adds Sign API framework, and a simple example client to demonstrate pairing and session creation capabilities.

This change introduces the following functionality:
- Sign API payload serializable/deserializable data types
- Payload encryption and decryption
- Session Diffie–Hellman key derivation
- Pairing URI parsing
- Simple WS based client

Resolves #47

## How Has This Been Tested?

- [x] Use example client to pair and establish the session with an example (DApp)[https://react-app.walletconnect.com/]:
      - Click on "Goerli" and then "connect"
      - In a new pop-up, click "New Pairing"
      - Copy the Pairing URI
      - In the terminal, `cd path/to/WalletConnectRust`
      - `.../WalletConnectRust$ cargo run --example session "<copied URI>"`
      - DApp should now display the session window
      - In the DApp click disconnect

![image](https://github.com/silvestrst-crypto/WalletConnectRust/assets/103181555/24d1bc4c-e11c-4723-be41-00c9c62788be)


- [x] Unittests

## Due Diligence

* [ ] Breaking change
* [x] Requires a documentation update
* [ ] Requires a e2e/integration test update

## TODO

- [ ] Typed errors instead of unnamed anyhow errors
- [ ] More unittests
- [ ] Better documentation, code comments and fuction descriptions
- [x] `crypto/session.rs` refactor

## Additional information

WCv2 uses layered approach. The relay protocol messages (currently only IRN, but there might be other in future), are used to carry the useful payload. In terms of this change the useful payload is Sign API messages.

Useful materials:
https://specs.walletconnect.com/2.0/specs/clients/core/pairing
https://specs.walletconnect.com/2.0/specs/clients/sign/session-proposal

https://specs.walletconnect.com/2.0/specs/servers/relay/relay-server-rpc
https://specs.walletconnect.com/2.0/specs/clients/sign/rpc-methods

silvestrst-crypto commented on sign_api/src/rpc.rs:
```diff
@@ -0,0 +1,138 @@
+}
+
+impl Payload {
+    /// Returns the message ID contained within the payload.
+    pub fn id(&self) -> u64 {
+        match self {
+            Self::Request(req) => req.id,
+            Self::Response(res) => res.id,
+        }
+    }
```
Not really used, can be removed?

silvestrst-crypto commented on sign_api/src/rpc.rs:
```diff
@@ -0,0 +1,138 @@
+//! The crate exports common types used when interacting with messages between
+//! clients. This also includes communication over HTTP between relays.
```
Bad Copy/Paste, should be changed to an appropriate description.

silvestrst-crypto commented on sign_api/src/rpc.rs:
```diff
@@ -0,0 +1,138 @@
+        }
+    }
+
+    pub fn validate(&self) -> Result<(), ValidationError> {
+        match self {
+            Self::Request(request) => request.validate(),
+            Self::Response(response) => response.validate(),
+        }
+    }
```
Should add more request/response specific validations.

silvestrst-crypto commented on sign_api/src/rpc/params/shared_types.rs:
```diff
@@ -0,0 +1,132 @@
+#[derive(Debug, Serialize, PartialEq, Eq, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct Namespace {
+    pub chains: Vec<String>,
+    pub methods: Vec<String>,
+    pub events: Vec<String>,
```
Might be cleaner to use `HashSet`, checking whether `Namespace` implements all required elements, potentially becomes a single function call.

silvestrst-crypto replied:
Done.

silvestrst-crypto commented:
@heilhead , apologies for names dropping! The repository doesn't look very active, so this is my attempt to get some attention to the PR. Would you please be able to give this change a quick review, or maybe assign someone? Thanks!

heilhead commented:
Hi @silvestrst-crypto, thanks for the PR! Sorry we didn't see it earlier. We'll try to review it in the next few days hopefully.

silvestrst-crypto commented:
gentle ping

heilhead commented:
@silvestrst-crypto Hey! Apologies for not getting back with the review, as I'm on holidays right now.

@xDarksome @farazdagi @chris13524 Can one of you guys please have a look at this?

farazdagi commented on sign_api/src/crypto/payload.rs:
```diff
@@ -0,0 +1,238 @@
+const TYPE_1: u8 = 1;
+const TYPE_INDEX: usize = 0;
+const TYPE_LENGTH: usize = 1;
+const IV_LENGTH: usize = 12;
```
I'd probably name it a bit more explicitly, sth like `INIT_VEC_LEN` (`IV` looks like Roman digit 4).

NB: Everywhere `iv` is used, I'd probably have it as `InitVecFoo`, `INIT_VEC`, `init_vec` etc.

farazdagi commented on sign_api/src/rpc/params/session_settle.rs:
```diff
@@ -0,0 +1,57 @@
+pub struct SessionSettleRequest {
+    pub relay: Relay,
+    pub controller: Controller,
+    pub namespaces: SettleNamespaces,
```
The optional `specialNamespaces` are omitted here. Not a big deal though.

farazdagi commented on sign_api/src/rpc/params/session_settle.rs:
```diff
@@ -0,0 +1,57 @@
+    pub relay: Relay,
+    pub controller: Controller,
+    pub namespaces: SettleNamespaces,
+    /// uSecs contrary to what documentation says (secs).
+    pub expiry: u64,
```
I've requested more info from the team on this -- if microseconds are used indeed, we need to update the documentation.

silvestrst-crypto replied:
When I bump down the expiry to 300 (5 min) in seconds, then I get an expiry error on `session_settle`:
![image](https://github.com/WalletConnect/WalletConnectRust/assets/103181555/0b51fcf3-37a2-4d87-b6e2-3c837a10c1b7)

silvestrst-crypto replied:
When I have poked around it a long time ago, I used example wallet and put some debug statements to see the raw data passed around, and this was the example payload for `session_settle`:
![image](https://github.com/WalletConnect/WalletConnectRust/assets/103181555/d37aa8c8-5fad-4e11-bdac-d05ee882c825)

silvestrst-crypto replied:
Which makes is almost 30min in usec.

farazdagi replied:
Thank you for extra info -- will make sure I follow this up with API team to understand whether the issue is still there.

ganchoradkov replied:
Hey @silvestrst-crypto, chiming in from JS team. That `session` expiry is indeed in `milliseconds` while all? other `TTL`/user-defined expirires are in seconds. I will get the team to update the docs, thank you 💯

silvestrst-crypto replied:
Are you guys sure its not `uSec`? I have just changed `uSec` to `ms` in my session example:
```
fn create_settle_request(responder_public_key: String) -> RequestParams {
    RequestParams::SessionSettle(SessionSettleRequest {
        relay: Relay {
            protocol: SUPPORTED_PROTOCOL.to_string(),
            data: None,
        },
        controller: Controller {
            public_key: responder_public_key.to_string(),
            metadata: Metadata {
                name: format!("Rust session example: {}", Utc::now()),
                icons: vec!["https://www.rust-lang.org/static/images/rust-logo-blk.svg".to_string()],
                ..Default::default()
            },
        },
        namespaces: supported_settle_namespaces(supported_account()),
        expiry: 300000, // 5 min in ms
    })
}
```

And it is expiring still with the same message as in the screenshot above.

silvestrst-crypto replied:
Worked, must be the same concept as for the `wc_sessionrequest`, just wasn't explicitly mentioned in the docs.

arein commented:
@silvestrst-crypto can you please also share your use case of how you're intending of using this? 🙏

farazdagi reviewed (commented):
Absolutely marvellous work, thank you!
I will do another review pass today -- but overall looks really great. I had some minor comments.

NB: Sorry for keeping you waiting for review for so long, will make sure we progress swiftly from this point on :)

farazdagi commented:
From your todo list, I'd consider typed errors + Readme update + probably a bit more unit tests for this PR. Anything else, let's tackle as separate PRs.

![image](https://github.com/WalletConnect/WalletConnectRust/assets/188194/9e31b537-1191-46ae-8d89-114863ecfa34)

silvestrst-crypto commented:
Thank you guys for your reviews! I will hopefully aim to go through the comments properly tomorrow and fix them up.

silvestrst-crypto commented:
> @silvestrst-crypto can you please also share your use case of how you're intending of using this? 🙏

We have an internal Rust project, where we are planning on using WalletConnect v2 to do DeFI operations, which is the main reason for tackling the Sign API at this point :smile:

silvestrst-crypto reviewed (commented):
Thank you guys for your reviews!

This is the first pass where I have tackled some of the review comments, I will handle the rest in next few passes.

Summary:
- Switches to typed errors rather than `anyhow::Error`
- Changed `Namespace` memebrs from `Vec` to `BTreeSet`
- Fixed-up some naming comments in `payload.rs`

I have also realised that `Namespace` needs more validation logic in accordance to the spec.

When you got through comments, could you guys only resolve those that you are happy with, but keep open those that you think are still "not there yet". This should make the review process a bit easier for both parties. Thanks!

silvestrst-crypto reviewed (commented):
This is a second part of the review comments addressing:
- Fixed `SessionRequestRequest` to match the spec (was missing an optional field, and `params` were changed to `serde_json::Value`)
- Fixed  `SessionEventRequest` to match the spec (`data` was changed to `serde_json::Value`)
- Simple serealize/deserialize unittests were added for the above two bullet-points
- Some other minor comment were fixed

chris13524 commented on examples/session.rs:
```diff
@@ -0,0 +1,522 @@
+//! Simple WalletConnect v2 client to demonstrate Sign API functionality.
```
Can we move this example to `sign_api` folder?

silvestrst-crypto replied:
Done.

silvestrst-crypto commented:
> First of all, thanks for the contribution! We really appreciate it 🥇
> 
> Second, can you add a section to the root README for the `sign_api` crate? We also need to add a warning to that; the Rust team doesn't work on the client SDKs so the amount of rigor here is far less than e.g. our JavaScript, Kotlin, or Swift SDKs have gone through. Those SDKs are also continually improved with new features and fixes, so the Rust version will likely not be getting any of these updates unless a community member contributes it. This could change in the future if there is sufficient interest in the Rust SDK, but for now we cannot dedicate resources to keep this maintained. Can you add a note to the readme like:
> 
> > Warning: this Rust Sign API SDK is community-maintained and may be lacking features and stability or security fixes that other versions of the Sign API SDK receive. We strongly recommend using the JavaScript or other Sign API SDKs instead.

Good idea, I understand, thanks!

silvestrst-crypto commented on sign_api/src/rpc/params/shared_types.rs:
```diff
@@ -0,0 +1,174 @@
+/// https://specs.walletconnect.com/2.0/specs/clients/sign/namespaces
+#[derive(Debug, Serialize, Eq, PartialEq, Deserialize, Clone, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct Namespaces {
```
```
#[derive(Debug, Serialize, Eq, PartialEq, Deserialize, Clone, Default)]
#[serde(rename_all = "camelCase")]
pub struct Namespaces(BTreeMap<String, Namespace>);
```
Makes it more flexible, easier to validate, and can match CAIP-2 field names.
https://specs.walletconnect.com/2.0/specs/clients/sign/namespaces

silvestrst-crypto replied:
Done.

silvestrst-crypto commented on sign_api/src/rpc/params/session_event.rs:
```diff
@@ -0,0 +1,68 @@
+        let deserialized: SessionEventRequest = serde_json::from_str(&expected)?;
+        let actual = serde_json::to_string(&deserialized)?;
+
+        assert_eq!(expected, actual);
```
Done.

silvestrst-crypto commented:
In the latest push I have:
- added simple serialization/deserialization unittests for the parameters
- done significant `ProposeNamespaces` refactor and added unittests
- renamed `Namespaces` to `ProposeNamespaces` and moved them together with `SettleNamespaces` under their own sub-foleders in `shred_types`
- Fixed up TTL for `session_request`, it's `unix_timestamp_in_secs()` + TTL

silvestrst-crypto commented:
In the latest push:
- I have moved `session` example  into `sign_api` crate

silvestrst-crypto commented:
In the latest push:
- Added README.md to the `sign_api` crate
- Rebased my changes on upstream main

insipx commented:
Is there a reason a JSON-RPC library supporting websockets cannot be used for implementing the Client? (for instance, [`jsonrpsee`](https://crates.io/crates/jsonrpsee)). That might make the code a little cleaner, as well as potentially solve some of the shortfalls:

```
//! Please note that this client is for demonstration purposes only, and hence
//! it doesn't handle many things that a complete client should handle:
//! - TTL control
//! - Out-of-order requests/responses (potentially)
//! - proper session and pairing control
//! - ...
```

nopestack commented:
Hi @silvestrst-crypto, thank you for your hard work on this PR 💪 . I was wondering if you had made any progress addressing the feedback from my colleagues and if you'd like to continue working on this issue?

silvestrst-crypto commented:
Hi @nopestack , lately I am quite busy, so put it on hold, but if anyone is planning on using it, I can slowly start moving it forward in spare time. Thanks!

dougEfresh commented:
@nopestack I think it is best to move this code to an isolated repo, and depend on wallet-connect-rust . 
I think this would speed up development and I'm willing to help. Besides, I do not think the maintainers have interest in merging this.

nopestack commented:
Marking this as draft in the meantime as work is paused and we cannot prioritize it at the moment. Worth revisiting it later @xDarksome @chris13524
